File: src/main/java/org/veupathdb/lib/gradle/container/config/RedirectConfig.java
package org.veupathdb.lib.gradle.container.config;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.util.Objects;

@SuppressWarnings("unused")
public class RedirectConfig {
  private static final byte ToStdOut  = 0;
  private static final byte ToStdErr  = 1;
  private static final byte ToDevNull = 2;
  private static final byte ToFile    = 3;
  private static final byte ToWriter  = 4;
  private static final byte ToPrint   = 5;
  private static final byte ToStream  = 6;

  private final Object value;

  private final byte type;

  private RedirectConfig(byte type) {
    this.value = null;
    this.type  = type;
  }

  private RedirectConfig(@NotNull File file) {
    type  = ToFile;
    value = file;
  }

  private RedirectConfig(@NotNull PrintStream writer) {
    type  = ToPrint;
    value = writer;
  }

  private RedirectConfig(@NotNull Writer writer) {
    type  = ToWriter;
    value = writer;
  }

  private RedirectConfig(@NotNull OutputStream writer) {
    type  = ToStream;
    value = writer;
  }

  @Contract(pure = true)
  public void forStdOut(@NotNull Process proc) {
    apply(proc.getInputStream());
  }

  @Contract(pure = true)
  public void forStdErr(@NotNull Process proc) {
    apply(proc.getErrorStream());
  }

  @Contract(pure = true)
  @SuppressWarnings("ConstantConditions")
  private void apply(@NotNull InputStream in) {
    switch (type) {
      case ToStdOut -> applyStd(in, System.out);
      case ToStdErr -> applyStd(in, System.err);
      case ToDevNull -> pipe(in, new DiscardingOutputStream());
      case ToFile -> applyFile(in, (File) value);
      case ToWriter -> applyWriter(in, (Writer) value);
      case ToPrint -> applyStd(in, (PrintStream) value);
      case ToStream -> pipe(in, (OutputStream) value);
    }
  }

  @Contract(pure = true)
  private void applyStd(@NotNull InputStream stream, @NotNull PrintStream tgt) {
    final var buf = stream instanceof BufferedInputStream
      ? (BufferedInputStream) stream
      : new BufferedInputStream(stream);

    pipe(buf, new BufferedOutputStream(tgt));
  }

  @Contract(pure = true)
  private void applyFile(@NotNull InputStream in, @NotNull File out) {
    try {
      if (!out.exists()) {
        //noinspection ResultOfMethodCallIgnored
        out.createNewFile();
      }

      pipe(in, new FileOutputStream(out));
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Contract(pure = true)
  private void applyWriter(@NotNull InputStream in, @NotNull Writer out) {
    pipe(in, new WriterOutputStream(out));
  }

  @Contract(pure = true)
  private void pipe(@NotNull InputStream in, @NotNull OutputStream out) {
    var input = in instanceof BufferedInputStream
      ? (BufferedInputStream) in
      : new BufferedInputStream(in);

    var output = out instanceof BufferedOutputStream
      ? (BufferedOutputStream) out
      : new BufferedOutputStream(out);

    pipe(input, output);
  }

  @Contract(pure = true)
  private void pipe(@NotNull BufferedInputStream in, @NotNull BufferedOutputStream out) {

    new Thread(() -> {
      final var bytes = new byte[8192];

      try {
        var read = 0;
        while ((read = in.read(bytes)) > -1) {
          out.write(bytes, 0, read);
          out.flush();
        }
      } catch (IOException e) {
        throw new RuntimeException(e);
      } finally {
        try {
          out.close();
        } catch (IOException e) {
          //noinspection ThrowFromFinallyBlock
          throw new RuntimeException(e);
        }
      }
    }).start();
  }

  //
  //
  // Static Constructors
  //
  //


  @NotNull
  public static RedirectConfig toStdOut() {
    return new RedirectConfig(ToStdOut);
  }

  @NotNull
  public static RedirectConfig toStdErr() {
    return new RedirectConfig(ToStdErr);
  }

  @NotNull
  public static RedirectConfig toDevNull() {
    return new RedirectConfig(ToDevNull);
  }

  @NotNull
  public static RedirectConfig toFile(@NotNull String file) {
    return new RedirectConfig(new File(Objects.requireNonNull(file)));
  }

  @NotNull
  public static RedirectConfig toFile(@NotNull File file) {
    return new RedirectConfig(Objects.requireNonNull(file));
  }

  @NotNull
  public static RedirectConfig toWriter(@NotNull Writer writer) {
    return new RedirectConfig(Objects.requireNonNull(writer));
  }

  @NotNull
  public static RedirectConfig toPrintStream(@NotNull PrintStream stream) {
    return new RedirectConfig(Objects.requireNonNull(stream));
  }

  @NotNull
  public static RedirectConfig toStream(@NotNull OutputStream stream) {
    return new RedirectConfig(Objects.requireNonNull(stream));
  }

  //
  //
  // Helper Classes
  //
  //

  private static class WriterOutputStream extends OutputStream {
    private final Writer writer;

    public WriterOutputStream(Writer writer) {
      this.writer = writer;
    }

    @Override
    public void write(int b) throws IOException {
      writer.write(b);
    }
  }

  private static class DiscardingOutputStream extends OutputStream {
    @Override
    public void write(int b) {
      // Do nothing.
    }

    @Override
    @SuppressWarnings("NullableProblems")
    public void write(@NotNull byte[] b) {
      // Do nothing.
    }

    @Override
    @SuppressWarnings("NullableProblems")
    public void write(@NotNull byte[] b, int off, int len) {
      // Do nothing.
    }

    @Override
    public void flush() {
      // Do nothing.
    }

    @Override
    public void close() {
      // Do nothing.
    }
  }
}

File: src/main/java/org/veupathdb/lib/gradle/container/tasks/base/exec/RedirectConfig.kt
package org.veupathdb.lib.gradle.container.tasks.base.exec

import java.io.*
import java.util.Objects

@SuppressWarnings("unused")
class RedirectConfig {
  companion object {
    const val ToStdOut:  Byte = 0
    const val ToStdErr:  Byte = 1
    const val ToDevNull: Byte = 2
    const val ToFile:    Byte = 3
    const val ToWriter:  Byte = 4
    const val ToPrint:   Byte = 5
    const val ToStream:  Byte = 6

    @JvmStatic
    fun toStdOut() = RedirectConfig(ToStdOut)

    @JvmStatic
    fun toStdErr() = RedirectConfig(ToStdErr)

    @JvmStatic
    fun toDevNull() = RedirectConfig(ToDevNull)

    @JvmStatic
    fun toFile(file: String) = RedirectConfig(File(file))

    @JvmStatic
    fun toFile(file: File) = RedirectConfig(Objects.requireNonNull(file))

    @JvmStatic
    fun toWriter(writer: Writer) = RedirectConfig(writer)

    @JvmStatic
    fun toPrintStream(stream: PrintStream) = RedirectConfig(stream)

    @JvmStatic
    fun toStream(stream: OutputStream) = RedirectConfig(stream)
  }

  private val value: Any?

  private val type: Byte

  private constructor(type: Byte) {
    this.value = null
    this.type  = type
  }

  private constructor(file: File) {
    type  = ToFile
    value = file
  }

  private constructor(writer: PrintStream) {
    type  = ToPrint
    value = writer
  }

  private constructor(writer: Writer) {
    type  = ToWriter
    value = writer
  }

  private constructor(writer: OutputStream) {
    type  = ToStream
    value = writer
  }

  fun forStdOut(proc: Process) = apply(proc.inputStream)

  fun forStdErr(proc: Process) = apply(proc.errorStream)

  fun apply(`in`: InputStream) {
    when (type) {
      ToStdOut -> applyStd(`in`, System.out)
      ToStdErr -> applyStd(`in`, System.err)
      ToDevNull -> pipe(`in`, DiscardingOutputStream())
      ToFile -> applyFile(`in`, value as File)
      ToWriter -> applyWriter(`in`, value as Writer)
      ToPrint -> applyStd(`in`, value as PrintStream)
      ToStream -> pipe(`in`, value as OutputStream)
    }
  }

  private fun applyStd(stream: InputStream, tgt: PrintStream) {
    val buf = when(stream) {
      is BufferedInputStream -> stream
      else                   -> BufferedInputStream(stream)
    }

    pipe(buf, BufferedOutputStream(tgt))
  }

  fun applyFile(i: InputStream, o: File) {
    if (!o.exists()) {
      //noinspection ResultOfMethodCallIgnored
      o.createNewFile()
    }

    pipe(i, FileOutputStream(o))
  }

  fun applyWriter(i: InputStream, o: Writer) = pipe(i, WriterOutputStream(o))

  fun pipe(i: InputStream, o: OutputStream) {
    val input = when(i) {
      is BufferedInputStream -> i
      else                   -> BufferedInputStream(i)
    }

    val output = when(o) {
      is BufferedOutputStream -> o
      else                    -> BufferedOutputStream(o)
    }

    pipe(input, output)
  }

  fun pipe(i: BufferedInputStream, o: BufferedOutputStream) {
    Thread {
      val bytes = ByteArray(8192)
      var read = i.read(bytes)

      while (read > -1) {
        o.write(bytes, 0, read)
        o.flush()
        read = i.read(bytes)
      }
    }.start()
  }

  //
  //
  // Helper Classes
  //
  //

  private class WriterOutputStream(val writer: Writer): OutputStream() {
    override fun write(b: Int) { writer.write(b) }
  }

  private class DiscardingOutputStream: OutputStream() {
    override fun write(b: Int) {
      // Do nothing.
    }

    override fun write(b: ByteArray) {
      // Do nothing.
    }

    override fun write(b: ByteArray, off: Int, len: Int) {
      // Do nothing.
    }

    override fun flush() {
      // Do nothing.
    }

    override fun close() {
      // Do nothing.
    }
  }
}
